x-pack/libbeat/autodiscover/providers/aws/ec2/fetch.go

```diff
@@ -0,0 +1,230 @@
+// Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+// or more contributor license agreements. Licensed under the Elastic License;
+// you may not use this file except in compliance with the Elastic License.
+
+package ec2
+
+import (
+	"context"
+	"sync"
+
+	"github.com/aws/aws-sdk-go-v2/service/ec2"
+	"github.com/aws/aws-sdk-go-v2/service/ec2/ec2iface"
+	"go.uber.org/multierr"
+
+	"github.com/elastic/beats/libbeat/logp"
+	awsauto "github.com/elastic/beats/x-pack/libbeat/autodiscover/providers/aws"
+)
+
+// fetcher is an interface that can fetch a list of ec2Instance objects without pagination being necessary.
+type fetcher interface {
+	fetch(ctx context.Context) ([]*ec2Instance, error)
+}
+
+// apiMultiFetcher fetches results from multiple clients concatenating their results together
+// Useful since we have a fetcher per region, this combines them.
+type apiMultiFetcher struct {
+	fetchers []fetcher
+}
+
+func (amf *apiMultiFetcher) fetch(ctx context.Context) ([]*ec2Instance, error) {
+	fetchResults := make(chan []*ec2Instance)
+	fetchErr := make(chan error)
+
+	// Simultaneously fetch all from each region
+	for _, f := range amf.fetchers {
+		go func(f fetcher) {
+			res, err := f.fetch(ctx)
+			if err != nil {
+				fetchErr <- err
+			} else {
+				fetchResults <- res
+			}
+		}(f)
+	}
+
+	var results []*ec2Instance
+	var errs []error
+
+	for pending := len(amf.fetchers); pending > 0; pending-- {
```

Comment:
nice :grimacing: It took me a while to wrap my head around this method but now I see it's simple and elegant and better than to use wait groups and closing channels

Reply:
I copied it from elb provider 😬 cool code!